lib/crawler/data/extraction/rule.rb

```diff
@@ -0,0 +1,92 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+
+# frozen_string_literal: true
+
+module Crawler
+  module Data
+    module Extraction
+      class Rule
+        ACTION_TYPE_EXTRACT = 'extract'
+        ACTION_TYPE_SET = 'set'
+        ACTIONS = [ACTION_TYPE_EXTRACT, ACTION_TYPE_SET].freeze
+        JOINS = %w[array string].freeze
+        SOURCES = %w[url html].freeze
+        RESERVED_FIELD_NAMES = %w[body_content domains headings meta_description title url url_host url_path
```

Comment:
This feels like an odd place for this constant.

Reply:
I moved this to a new module, under `lib/constants.rb`. I've never made a const-only module before but it apparently isn't unusual.
There's a few other consts I could see moving here so they're easier to reference and configure, but I will look at that in another PR.